sync/syncer.go

```diff
@@ -485,27 +485,41 @@ func (s *Syncer) DataAvailabilty(blk *types.Block) ([]*types.AddressableSignedTr
 	return txs, atxs, nil
 }
 
-func (s *Syncer) fetchLayerBlockIds(m map[string]p2p.Peer, lyr types.LayerID) ([]types.BlockID, error) {
+func (s *Syncer) fetchLayerBlockIds(m map[uint32][]p2p.Peer, lyr types.LayerID) ([]types.BlockID, error) {
 	//send request to different users according to returned hashes
-	v := make([]p2p.Peer, 0, len(m))
-	for _, value := range m {
-		v = append(v, value)
-	}
-
-	wrk, output := NewPeersWorker(s, v, &sync.Once{}, LayerIdsReqFactory(lyr))
-	go wrk.Work()
-
 	idSet := make(map[types.BlockID]struct{}, s.LayerSize)
 	ids := make([]types.BlockID, 0, s.LayerSize)
+	for h, peers := range m {
+	NextHash:
+		for _, peer := range peers {
+			s.Info("send request Peer: %v", peer)
+			ch, err := LayerIdsReqFactory(lyr)(s.MessageServer, peer)
+			if err != nil {
+				return nil, err
+			}
 
-	//unify results
-	for out := range output {
-		if out != nil {
-			//filter double ids
-			for _, bid := range out.([]types.BlockID) {
-				if _, exists := idSet[bid]; !exists {
-					idSet[bid] = struct{}{}
-					ids = append(ids, bid)
+			timeout := time.After(s.RequestTimeout)
+			select {
+			case <-timeout:
```

Comment:
To my understanding, values may not be written to the channel (on error).
If that is the case, a better way to do it is by writing nil to ch - then we will not waste time waiting for the timeout.

Reply:
@almogdepaz ?

Reply:
the channel is closed on error